packages/loader/container-loader/src/prefetchDocumentStorageService.ts

```diff
@@ -15,10 +18,9 @@ export class PrefetchDocumentStorageService extends DocumentStorageServiceProxy
     private prefetchEnabled = true;
 
     public async getSnapshotTree(version?: IVersion): Promise<ISnapshotTree | null> {
-        const p = this.internalStorageService.getSnapshotTree(version);
+        const p = readWithRetry(async () => this.internalStorageService.getSnapshotTree(version));
```

Comment:
I do not think this is the right layer to add retries. This layer is not even created in case we use SPO, because this layer responsibility is to prefetch, and nothing else, and SPO already does prefetching, so we do not create this object.

If you want to go with similar workflow, then it has to be its own adapter class that implements retries, but has no other responsibilities, and can be created by Container.
I'll only add that Container has to be aware of it and let it know that container is cancelled, such that any retrys can be canceled.

Reply:
We do create this for all, in case of spo we just call stopPrefetch which disables the prefetching. But I am looking for  your suggestions to add it as another layer and keep separation of concerns.

Reply:
Please see https://github.com/microsoft/FluidFramework/pull/4051 and corresponding main commit. This code is optional, it has only one purpose - prefetch. We should not put other responsibilities into this layer

Reply:
Added another layer as retriable storage service.